cnd/src/btsieve/bitcoin/mod.rs

```diff
@@ -1,152 +1,217 @@
 };
 use chrono::NaiveDateTime;
 use futures_core::compat::Future01CompatExt;
+use genawaiter::{
+    sync::{Co, Gen},
+    GeneratorState,
+};
 use reqwest::{Client, Url};
 use std::collections::HashSet;
 
-pub async fn matching_transaction<C>(
-    mut blockchain_connector: C,
-    pattern: TransactionPattern,
+type Hash = bitcoin::BlockHash;
+type Block = bitcoin::Block;
+
+pub async fn watch_for_spent_outpoint<C>(
+    blockchain_connector: C,
     start_of_swap: NaiveDateTime,
-) -> anyhow::Result<transaction::Bitcoin>
+    from_outpoint: OutPoint,
+    unlock_script: Vec<Vec<u8>>,
+) -> anyhow::Result<bitcoin::Transaction>
 where
-    C: LatestBlock<Block = bitcoin::Block>
-        + BlockByHash<Block = bitcoin::Block, BlockHash = bitcoin::BlockHash>
-        + Clone,
+    C: LatestBlock<Block = Block> + BlockByHash<Block = Block, BlockHash = Hash> + Clone,
 {
-    // Verify that we can successfully addPeer to the blockchain connector and check
-    // if the transaction is in the latest block.
-    let latest_block = blockchain_connector.latest_block().compat().await?;
-    if let Some(transaction) = check_block_against_pattern(&latest_block.clone(), &pattern) {
-        return Ok(transaction.clone());
-    };
-
-    // We didn't find the transaction, now we need to do two things; keep polling
-    // for latest block so that we see transactions in new blocks and also go
-    // back up the blockchain until 'start_of_swap' i.e., look back in the
-    // past.
-
-    let mut prev_blockhashes: HashSet<bitcoin::BlockHash> = HashSet::new();
-    let mut missing_block_futures: Vec<_> = Vec::new();
+    let transaction = matching_transaction(blockchain_connector, start_of_swap, |transaction| {
+        transaction.spends_from_with(&from_outpoint, &unlock_script)
+    })
+    .await?;
 
-    let mut oldest_block: Option<bitcoin::Block> = Some(latest_block.clone());
-    prev_blockhashes.insert(latest_block.bitcoin_hash());
+    Ok(transaction)
+}
 
-    let prev_blockhash = latest_block.header.prev_blockhash;
-    let future = blockchain_connector.block_by_hash(prev_blockhash).compat();
-    missing_block_futures.push((future, prev_blockhash));
+pub async fn watch_for_created_outpoint<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    compute_address: bitcoin::Address,
+) -> anyhow::Result<(bitcoin::Transaction, bitcoin::OutPoint)>
+where
+    C: LatestBlock<Block = Block> + BlockByHash<Block = Block, BlockHash = Hash> + Clone,
+{
+    let transaction = matching_transaction(blockchain_connector, start_of_swap, |transaction| {
+        transaction.spends_to(&compute_address)
+    })
+    .await?;
 
-    loop {
-        // Delay so that we don't overload the CPU in the event that
-        // latest_block() and block_by_hash() resolve quickly.
+    let (vout, _txout) = transaction
+        .find_output(&compute_address)
+        .expect("Deployment transaction must contain outpoint described in pattern");
```

Comment:
Stale string (refers to 'pattern').

Reply:
Good catch!
Luckily, that whole error path is going away soon :)